Lib/textwrap.py

```diff
@@ -429,42 +426,27 @@ def dedent(text):
 
     Entirely blank lines are normalized to a newline character.
     """
-    # Look for the longest leading string of spaces and tabs common to
-    # all lines.
-    margin = None
-    text = _whitespace_only_re.sub('', text)
-    indents = _leading_whitespace_re.findall(text)
-    for indent in indents:
-        if margin is None:
-            margin = indent
-
-        # Current line more deeply indented than previous winner:
-        # no change (previous winner is still on top).
-        elif indent.startswith(margin):
-            pass
-
-        # Current line consistent with and no deeper than previous winner:
-        # it's the new winner.
-        elif margin.startswith(indent):
-            margin = indent
-
-        # Find the largest common whitespace between current line and previous
-        # winner.
-        else:
-            for i, (x, y) in enumerate(zip(margin, indent)):
-                if x != y:
-                    margin = margin[:i]
-                    break
+    if not text:
+        return text
+
+    # If the input is entirely whitespace, return normalized lines
+    if text.isspace():
+        return '\n' * text.count('\n')
+
+    lines = text.split('\n')
 
-    # sanity check (testing/debugging only)
-    if 0 and margin:
-        for line in text.split("\n"):
-            assert not line or line.startswith(margin), \
-                   "line = %r, margin = %r" % (line, margin)
+    # Get length of leading whitespace, inspired by ``os.path.commonprefix()``
+    non_blank_lines = [l for l in lines if l and not l.isspace()]
+    l1 = min(non_blank_lines)
+    l2 = max(non_blank_lines)
+    margin = 0
+    for i, c in enumerate(l1):
+        if c != l2[i] or c not in ' \t':
+            break
+        margin += 1
 
-    if margin:
-        text = re.sub(r'(?m)^' + margin, '', text)
-    return text
+    return '\n'.join([l[margin:] if l and not l.isspace() else ''
+                      for l in lines])
```

Comment:
```suggestion
    return '\n'.join(l[margin:] if l and not l.isspace() else ''
                     for l in lines)
```

No need to materialize the list in memory.

Reply:
I did this because of PEP-0709, as I don't believe generators like this are inlined yet?

Reply:
Using the list comprehension is indeed faster (but will use more memory). Either option is fine with me.